### PR TITLE
fix: IRSA support for Kubernetes 1.24 described in issue #1177

### LINKED ISSUE #1177: [ENHANCEMENT] Kubernetes version 1.24 support
Support for version [1.24](https://aws.amazon.com/blogs/containers/amazon-eks-now-supports-kubernetes-version-1-24/)

### PR DESCRIPTION
**Pull request by guilhermeblanco**
### What does this PR do?

This addresses the issue on upgraded clusters of EKS version 1.24 returning the following warning when executing:

 Warning: "default_secret_name" is no longer applicable for Kubernetes v1.24.0 and above
│ 
│   with module.eks_addons.module.external_dns[0].module.helm_addon.module.irsa[0].kubernetes_service_account_v1.irsa[0],
│   on .terraform/modules/eks_addons/modules/irsa/main.tf line 16, in resource "kubernetes_service_account_v1" "irsa":
│   16: resource "kubernetes_service_account_v1" "irsa" {
│ 
│ Starting from version 1.24.0 Kubernetes does not automatically generate a token for service accounts, in this case, "default_secret_name" will be empty

### Motivation

- Resolves #1177

### More

- [ ] Yes, I have tested the PR using my local account setup (Provide any test evidence report under Additional Notes)
- [ ] Yes, I have added a new example under [examples](https://github.com/aws-ia/terraform-aws-eks-blueprints/tree/main/examples) to support my PR
- [ ] Yes, I have created another PR for add-ons under [add-ons](https://github.com/aws-samples/eks-blueprints-add-ons) repo (if applicable)
- [ ] Yes, I have updated the [docs](https://github.com/aws-ia/terraform-aws-eks-blueprints/tree/main/docs) for this feature
- [ ] Yes, I ran `pre-commit run -a` with this PR

**Note**: Not all the PRs require a new example and/or doc page. In general:
- Use an existing example when possible to demonstrate a new addons usage
- A new docs page under `docs/add-ons/*` is required for new a new addon

### For Moderators

- [ ] E2E Test successfully complete before merge?

### CODE REVIEW

**Comment by tanvp112** [comment]
The warning message is intended, see [here](https://github.com/hashicorp/terraform-provider-kubernetes/pull/1792) for details. Suggest blueprints should leave the secret creation to the user and not create it by default, the same goes to setting `automount_service_account_token`, there's no need to change this flag for 1.24.

**Comment by sabinayakc** [comment]
Yeah, I am facing the same issue after using 1.24.

**Comment by FernandoMiguel** [comment]
@guilhermeblanco would you mind syncing your branch with main?

File: modules/irsa/main.tf [hunk 1]
```diff
@@ -28,7 +44,7 @@ resource "kubernetes_service_account_v1" "irsa" {
     }
   }
 
-  automount_service_account_token = true
+  automount_service_account_token = false
```
Comment (pedroparraortega) [inline modules/irsa/main.tf:47]: If you set it to false, kubelet won't mount the credentials into the pod automatically causing the pod to fail.

File: modules/irsa/main.tf [hunk 2]
```diff
@@ -13,6 +13,22 @@ resource "kubernetes_namespace_v1" "irsa" {
...
+    namespace   = try(kubernetes_namespace_v1.irsa[0].metadata[0].name, var.kubernetes_namespace)
+    annotations = {
+      "kubernetes.io/service-account.name"      = var.kubernetes_service_account
+      "kubernetes.io/service-account.namespace" = try(kubernetes_namespace_v1.irsa[0].metadata[0].name, var.kubernetes_namespace)
+    }
+  }
+
+  type = "kubernetes.io/service-account-token"
+
+  depends_on = [kubernetes_service_account_v1.irsa]
```
Comment (Zvikan) [inline modules/irsa/main.tf:29]: Is this depends_on required? I'm going to run another test taking your code snippet here with our e2e tests , removing this depends on and keeping `automount_service_account_token` to true.

File: modules/irsa/main.tf [hunk 3]
```diff
@@ -13,6 +13,22 @@ resource "kubernetes_namespace_v1" "irsa" {
   }
 }
 
+resource "kubernetes_secret_v1" "irsa" {
+  count = var.create_kubernetes_service_account ? 1 : 0
+  metadata {
+    name        = format("%s-token-secret", var.kubernetes_service_account)
+    namespace   = try(kubernetes_namespace_v1.irsa[0].metadata[0].name, var.kubernetes_namespace)
+    annotations = {
+      "kubernetes.io/service-account.name"      = var.kubernetes_service_account
```
Comment (bryantbiggs) [inline modules/irsa/main.tf:22]: Change this and then drop the `depends_on`

```suggestion
      "kubernetes.io/service-account.name"      = try(kubernetes_service_account_v1.[0].metadata[0].name, var.kubernetes_service_account)
```

**Comment by FernandoMiguel** [comment]
we use the token in external secret operator to login to hashicorp vault
```hcl
resource "kubernetes_secret_v1" "external_secrets_token" {
  metadata {
    annotations = {
      "kubernetes.io/service-account.name" = var.service_account_name
    }
    namespace     = var.namespace
    generate_name = "external-secrets-token-"
  }

  type = "kubernetes.io/service-account-token"

  depends_on = [module.eks_blueprints_kubernetes_addons_external_secrets]
}
```

not sure about other use cases

**Pull request closed without merging**